### PR TITLE
Rootless Docker service

### PR DESCRIPTION
**Pull request by abbradar**
###### Motivation for this change

Docker with less security risks.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/virtualisation/docker-rootless.nix [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  ###### interface
+
+  options.virtualisation.docker.rootless = {
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        This option enables docker in a rootless mode, a daemon that manages
+        linux containers. To interact with the daemon, one needs to set
+        <command>DOCKER_HOST=unix://$XDG_RUNTIME_DIR/docker.sock</command>.
```
Comment (mikroskeem) [inline nixos/modules/virtualisation/docker-rootless.nix:22]: Perhaps set it by default via `environment.sessionVariables`, or by a separate option?

**Review by mikroskeem — APPROVED** [review]
Overall LGTM.

File: nixos/modules/virtualisation/docker-rootless.nix (hunk 1 above)
Comment (abbradar) [reply nixos/modules/virtualisation/docker-rootless.nix:22]: Good idea, I implemented that as `setSocketVariable` option, inspired by `ssh-agent` way of doing that.

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

**Review by mikroskeem — APPROVED** [review]

File: nixos/modules/virtualisation/docker-rootless.nix [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    setSocketVariable = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Point <command>DOCKER_HOST</command> to rootless Docker instance for
+        normal users by default.
+      '';
+    };
+
+    extraOptions = mkOption {
```
Comment (roberth) [inline nixos/modules/virtualisation/docker-rootless.nix:35]: This is not the best way to do it. See https://github.com/NixOS/nixpkgs/pull/141549

File: pkgs/applications/virtualization/docker/default.nix [hunk 3]
```diff
@@ -77,6 +77,8 @@ rec {
 
       extraPath = optionals (stdenv.isLinux) (makeBinPath [ iproute2 iptables e2fsprogs xz xfsprogs procps util-linux git ]);
 
+      extraUserPath = optionals (stdenv.isLinux) (makeBinPath [ rootlesskit slirp4netns fuse-overlayfs ]);
```
Comment (roberth) [inline pkgs/applications/virtualization/docker/default.nix:80]:
```suggestion
      extraUserPath = optionals (stdenv.isLinux && !clientOnly) (makeBinPath [ rootlesskit slirp4netns fuse-overlayfs ]);
```

Not the end of the world, but keeps the build closure lean(ish) for the client-only build.

File: nixos/modules/virtualisation/docker-rootless.nix (hunk 2 above)
Comment (abbradar) [reply nixos/modules/virtualisation/docker-rootless.nix:35]: Indeed, this way looks better. Given that the other PR seems almost ready, let's wait for it to land; then I can reuse the code.

**Comment by abbradar** [comment]
I updated the branch to use `daemon.settings`.

**Review by mikroskeem — APPROVED** [review]

**Comment by abbradar** [comment]
We have a test and several LGTMs, so let's merge.

**Pull request merged**